Adds a check to track extra js files and removes local compiled js before each compilation

ankita240796 opened the pull request:
## Explanation

- Adds a pre push check to ensure that if user has js files in diff, they should be present in JS_FILEPATHS_NOT_TO_BUILD
- Delete local_compiled_js before a new compilation to ensure old files are not present in the folder
- Adds a comment in tsconfig.json to explain extra included files

ankita240796 committed:
Remove local compiled js before each compilation

ankita240796 committed:
Fix pre push

ankita240796 committed:
Remove test file

ankita240796 committed:
Only js files in core and extensions should be checked

DubeySandeep commented on scripts/pre_push_hook.py:
```diff
@@ -295,21 +298,41 @@ def _install_hook():
+            extra_js_files.append(filename)
+
+    if len(extra_js_files):
+        err_msg = (
```
@ankita240796, I'm not sure whether it's a proper place to do this check. (As the name of the function doesn't indicate this functionality.)

ankita240796 replied:
Shall I add a new function which tests this behavior and is called by this function?

DubeySandeep replied:
I'm not sure whether we should do this kinda check in pre_push_hook.py, @apb7 Do you have any suggestion on this?

DubeySandeep replied:
Maybe we can create `does_diff_includes_frontend_scripts` which will check for .ts files and expected frontend .js files, if found returns true.
Whereas in pre_commit_linter.py we can put this test where we are testing whether the js file should be ts file.

I'm not sure about what patterns/layers we have for our linters. I'm asking this as I use to run pre_commit_linet before every commit so I expect all lint checks should be done in the pre_commit_linters. Leaving this to @apb7 to finalize the place for this test.

DubeySandeep replied:
@apb7, Ping!

apb7 replied:
I agree with @DubeySandeep here -- all checks should be placed in the pre-commit script. The general use of the pre-commit script should be, as the name suggests, to run before every commit. Thus, we need to include this check there.

>Maybe we can create does_diff_includes_frontend_scripts which will check for .ts files and expected frontend .js files, if found returns true.
Whereas in pre_commit_linter.py we can put this test where we are testing whether the js file should be ts file.

This sounds perfectly fine. Thanks!

DubeySandeep commented on scripts/pre_push_hook.py:
```diff
@@ -37,8 +37,11 @@
 
-# pylint: enable=wrong-import-order
+# pylint: disable=relative-import
+import build
```
Can we do `from . import xyz` and remove `pylint disable...`?

ankita240796 replied:
I tried that. It gives an error:

```
Traceback (most recent call last):
  File ".git/hooks/pre-push", line 40, in <module>
    from . import build
ValueError: Attempted relative import in non-package
```

DubeySandeep replied:
Thanks for checking!

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Fix names

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 commented:
Hi @apb7 @DubeySandeep can you PTAL, Thanks!

DubeySandeep commented:
@ankita240796, This PR LGTM, I'm waiting for @apb7 to reply on the above comment and review this PR! 

EDIT: Note, codecov test are failing!

ankita240796 committed:
Move test to pre commit script

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Update error message, remove extra code

ankita240796 commented:
Hi @DubeySandeep @apb7, I have moved the test to the pre-commit script, PTAL. Thanks!

ankita240796 committed:
Fix test

ankita240796 commented:
Hi @apb7, The e2e test for exploration history tab are failing on this PR eventhough no changes were made to any other files except the pre commit script. I am unable to understand why that happens. Can you please help me here, Thanks!

DubeySandeep commented:
Hi @ankita240796, I think it's happening because of #6587 (As the test passed on that PR ~15 days ago and we merged it without running the current tests.) I'm looking into it and I'll resolve the issue asap!

ankita240796 commented:
OK, Thanks @DubeySandeep!

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 commented:
Hi @DubeySandeep, can you please guide me about the failing codecov test. I am not sure about how to debug them. Thanks!

DubeySandeep commented:
@ankita240796, Thanks because your PR decreases the test-coverage by 0.01% (If you think you have added the test already, then try updating your branch with develop.)
![image](https://user-images.githubusercontent.com/16653571/57973173-4e8d2500-79c2-11e9-8463-86e613029ad5.png)

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Use defined functions, instead of if statements

ankita240796 commented:
Hi @DubeySandeep, thanks for the explanation. I have updated the code to use previously defined methods instead of new if else and no extra tests are now required, PTAL. Thanks!

DubeySandeep commented on scripts/pre_commit_linter.py:
```diff
@@ -2227,6 +2228,40 @@ def _check_codeowner_file(self):
+                filepath.endswith('.js'))]
+
+        extra_js_files = []
+        for filename in js_files_to_check:
```
@ankita240796, I think we redirect print to stdout (using _redirect_stdout).

Maybe we can have a simple and clean error message around the following pattern:
```
Starting extra js files check
----------------------------------------
a/b/c/d.js --> Found extra .js file
e/f/g/h.js --> Found extra .js file

If you want the above files to be present as js files, add them to the list JS_FILEPATHS_NOT_TO_BUILD in build.py else rename them to .ts
----------------------------------------
FAILED Extra js files check
``` 


(You can follow any lint check function in the pre_commit_linter file and see how the messages are printed.)

ankita240796 replied:
Done.

ankita240796 committed:
Merge branch 'fix-ts' of github.com:ankita240796/oppia into fix-ts

ankita240796 committed:
Add test

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 committed:
Update message printing

ankita240796 committed:
Fix lint

ankita240796 commented:
Hi @DubeySandeep, I have addressed the review comments as well as added new tests, PTAL. Thanks!

ankita240796 committed:
Fix test

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 commented:
Hi @DubeySandeep, I have added the test as well as updated my branch with develop. I am not sure why the tests are still failing. Can you PTAL, Thanks!

seanlip commented:
I don't think the tests are failing? The codecov/patch check is saying that there isn't enough test coverage in the PR.

You can see it in more detail by clicking the "Details" link on the right. The parts that are highlighted red are the parts that are not currently covered by tests.

ankita240796 commented:
Thanks, @seanlip! I will make updates to the test accordingly.

ankita240796 commented:
Hi @seanlip, the part which is not covered by test is an infinite while loop which runs to watch changes in tsc output log file when typescript is compiled in watch mode. The loop ends only when the tsc output log file contains a line `Watching for file changes`. I added a condition flag to ensure that the loop does not run during the test. 

There is no test for when the condition flag is true since that will make the loop to run infinitely and the tests will not end. [Here](https://codecov.io/gh/oppia/oppia/compare/76082ee41c122ac150445db83fb59850a1f428a9...bab48cb79dbdfe6ca53d6d3283aacb6bc26b8f3a/src/scripts/build.py#L1230) are the lines which are not covered.

I am not sure if we need a test for this case. If we need a test, shall I create a dummy tsc output log file and test with this file to ensure that loop ends as soon as the file contains the line `Watching for file changes`?

Thanks!

seanlip commented on scripts/build.py:
```diff
@@ -1181,32 +1181,42 @@ def generate_build_directory():
+        compiled_js_dir: str. The directory which contains the compiled js
+            files.
     """
+    safe_delete_directory_tree(compiled_js_dir)
```
It's a bit odd that the only thing this function is using compiled_js_dir is to delete it. How does it know where to put the files once they are compiled?

It would be better to have a single source of truth here for that dirpath. If tsc is getting it from somewhere else we should extract its value from there rather than passing it in.

Or you can declare COMPILED_JS_DIR as a global constant in this file and have a test that verifies that it matches the path specified in the source used by tsc. That might be better/cleaner.

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1215,7 +1225,7 @@ def compile_typescript_files_continuously(project_dir):
 
-    while True:
+    while condition:
         if os.path.isfile('tsc_output_log.txt'):
```
Hm, why would it not be a file ... I think you've just written to it a couple of lines earlier?

ankita240796 replied:
Yes, fixed this.

seanlip commented on scripts/build.py:
```diff
@@ -1215,7 +1225,7 @@ def compile_typescript_files_continuously(project_dir):
         subprocess.Popen(cmd, stdout=out)
 
-    while True:
+    while condition:
```
Generally we try not to have variables which say "we are running this in test mode". This can cause skew between the actual code and what the test is testing.

So, it would be better for this to be `while True` and to have a test that covers it. One approach you could take is to make tsc_output_log.txt a global constant in this file. You can then swap it out during tests to be whatever you want it to be, e.g. a dummy test file which has the trigger for termination.

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1181,32 +1181,42 @@ def generate_build_directory():
 
 
-def compile_typescript_files_continuously(project_dir):
+def compile_typescript_files_continuously(
```
I'm not sure what "continuously" means here, btw -- we seem to return once we see "Watching for file changes". Doesn't that mean this process terminates and is therefore not continuous by definition?

Or did you mean to say "when typescript is done with the initial compilation" in your last comment on lines 1234?

ankita240796 replied:
Yes, I mean to say that when typescript finishes initial compilation we can return from this function since the watcher is running in the background and the initial set of compiled files is also produced.

seanlip commented on scripts/build_test.py:
```diff
@@ -647,4 +648,35 @@ def test_generate_delete_tasks_to_remove_deleted_files(self):
+        compilation.
+        """
+        # pylint: disable=unused-argument
+        def mock_check_call(cmd):
```
You can replace cmd with unused_cmd, I think. That might remove the need for pylint-disable.

ankita240796 replied:
Done.

seanlip commented on scripts/build_test.py:
```diff
@@ -647,4 +648,35 @@ def test_generate_delete_tasks_to_remove_deleted_files(self):
+        # pylint: enable=unused-argument
+
+        compiled_js_dir = os.path.join(
+            'core', 'tests', 'build_sources', 'compiled_js_dir', '')
```
Ideally make use of the constants in build.py (see above comments).

ankita240796 replied:
Done.

seanlip reviewed (changes requested):
Hi @ankita240796, left some comments.  PTAL!

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Resolve conflicts

ankita240796 committed:
Address review comments

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

ankita240796 commented:
Hi @seanlip, I have addressed the review comments, PTAL. Thanks!

ankita240796 committed:
Fix test

seanlip commented on scripts/build.py:
```diff
@@ -112,6 +115,19 @@
+        out_dir = os.path.join(config_data['compilerOptions']['outDir'], '')
+    if out_dir != COMPILED_JS_DIR:
+        raise Exception(
+            'COMPILED_JS_DIR: %s is does not match the output directory '
```
is does --> does

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -112,6 +115,19 @@
 HASH_BLOCK_SIZE = 2**20
 
 
+def test_compiled_js_dir_is_valid():
```
Let's call this require_compiled_js_dir_to_be_valid() -- we only prefix stuff with test_ in the test files. 

Please add Raises docstring, too.

ankita240796 replied:
Done.

seanlip commented on scripts/build.py:
```diff
@@ -1190,6 +1206,8 @@ def compile_typescript_files(project_dir):
             to be compiled.
     """
+    test_compiled_js_dir_is_valid()
+    safe_delete_directory_tree(COMPILED_JS_DIR)
```
Just to check: if we delete the compiled js dir, does that result in any kind of slowdown? If so, regenerating everything might be an issue in "continuous run" mode, and I'm wondering if in that case there's a more fine-grained criterion we should use in order to figure out when to regenerate.

ankita240796 replied:
No, it does not. We are only deleting when we are doing a fresh compilation. When a watcher is running, there is no deletion but if the user shuts down the server and the compilation is triggered again, all files will be recompiled irrespective of whether the old files are present or not.

seanlip replied:
OK, sounds good. Thanks!

seanlip commented on scripts/build.py:
```diff
@@ -29,6 +29,9 @@
 ASSETS_OUT_DIR = os.path.join('build', 'assets', '')
 
+COMPILED_JS_DIR = os.path.join('local_compiled_js', '')
+TSC_OUTPUT_LOG_FILE = 'tsc_output_log.txt'
```
TSC_OUTPUT_LOG_FILEPATH

ankita240796 replied:
Done.

seanlip commented on scripts/build_test.py:
```diff
@@ -647,4 +653,58 @@ def test_generate_delete_tasks_to_remove_deleted_files(self):
+        """
+        def mock_check_call(unused_cmd):
+            pass
+        # pylint: disable=unused-argument
```
Just prefix the unused args with unused_, then you don't need this pylint pragma.

(I thought I mentioned this previously.)

ankita240796 replied:
Yes, I used unused_ with cmd but `shell` and `stdout` are passed as named args, so changing the name in mock results in failure of tests.

seanlip commented on scripts/build_test.py:
```diff
@@ -647,4 +653,58 @@ def test_generate_delete_tasks_to_remove_deleted_files(self):
+            build, 'COMPILED_JS_DIR', MOCK_COMPILED_JS_DIR), self.swap(
+                build, 'test_compiled_js_dir_is_valid',
+                mock_test_compiled_js_dir_is_valid):
+            if os.path.exists(os.path.dirname(MOCK_COMPILED_JS_DIR)):
```
This seems to be affirmatively deleting the compiled_js_dir, which invalidates the point of the test as described in the test name. Shouldn't we instead be ensuring that the dir exists (creating the dir if necessary), and then running the functions, and then verifying that the dir no longer exists?

See points 1 and 2 in [Writing Backend Tests](https://github.com/oppia/oppia/wiki/Writing-backend-tests#guidelines-for-writing-good-tests).

ankita240796 replied:
Updated.

seanlip commented on scripts/build.py:
```diff
@@ -1190,6 +1206,8 @@ def compile_typescript_files(project_dir):
+    safe_delete_directory_tree(COMPILED_JS_DIR)
     print 'Compiling ts files...'
     cmd = ['./node_modules/typescript/bin/tsc', '--project', project_dir]
     subprocess.check_call(cmd)
```
> Yes, I mean to say that when typescript finishes initial compilation we can return from this function since the watcher is running in the background and the initial set of compiled files is also produced.

Are you planning to change the docstring/comments to clarify?

ankita240796 replied:
The comment [here](url) mentions the point that `whenever typescript does the compilation, the last line reads Watching for file changes`. So, the update required here is to change compilation -> initial compilation?

seanlip replied:
Yes, please make that change. Thanks!

ankita240796 replied:
Done.

seanlip commented on scripts/build_test.py:
```diff
@@ -647,4 +653,58 @@ def test_generate_delete_tasks_to_remove_deleted_files(self):
+        build.test_compiled_js_dir_is_valid()
+
+        out_dir = ''
+        with open('tsconfig.json') as f:
```
Reuse the constant in build.py

ankita240796 replied:
Done.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2284,6 +2285,49 @@ def _check_codeowner_file(self):
+                filepath for filepath in self.all_filepaths if (
+                    filepath.endswith('.js'))]
+
+            for filename in js_files_to_check:
```
filename --> filepath (right)?

(Filename is just the last part, I think.)

ankita240796 replied:
Yes, done.

seanlip commented on scripts/pre_commit_linter.py:
```diff
@@ -2284,6 +2285,49 @@ def _check_codeowner_file(self):
+                err_msg = (
+                    'If you want the above files to be present as js files, '
+                    'add them to the list JS_FILEPATHS_NOT_TO_BUILD in '
+                    'build.py else rename them to .ts\n')
```
... build.py. Otherwise, rename ...

ankita240796 replied:
Done.

ankita240796 committed:
Address review comments, add comment in tsconfig for extra files

ankita240796 commented:
Hi @seanlip, I have addressed the review comments. I have also added a comment to explain extra files included in `tsconfig.json` as discussed in #6860. If you want a separate PR for that, I will revert them here. PTAL, Thanks!

seanlip reviewed (changes requested):
Looks good, just two small things left. Thanks @ankita240796!

seanlip commented:
Though it looks like the backend tests are failing...

ankita240796 committed:
Fix test

ankita240796 committed:
Update comment

ankita240796 committed:
Merge remote-tracking branch 'upstream/develop' into fix-ts

seanlip commented:
LGTM. Thanks @ankita240796!

@nithusha21 @apb7 PTAL as codeowners. This PR has been around for a while and we should try to get it merged. Thanks!

nithusha21 reviewed (approved):
LGTM as a codeowner

apb7 reviewed (approved):
LGTM, thanks @ankita240796!

DubeySandeep commented on core/tests/build_sources/mock_tsc_output_log.txt:
```diff
@@ -0,0 +1,4 @@
+c5:26:51 AM - Starting compilation in watch mode...
```
@ankita240796, is this an expected change? (Do we need this file?)

ankita240796 replied:
Yes, we need this for testing.

DubeySandeep replied:
(Ignore this comment)*
